xarray/backends/zarr.py

```diff
@@ -0,0 +1,474 @@
+    else:
+        for k in list(encoding):
+            if k not in valid_encodings:
+                del encoding[k]
+
+    chunks = _determine_zarr_chunks(encoding.get('chunks'), variable.chunks,
+                                    variable.ndim)
+    encoding['chunks'] = chunks
+
+    # TODO: figure out how to serialize compressor and filters options
+    # in zarr these are python objects, not strings
+
+    return encoding
+
+
+class ZarrStore(WritableCFDataStore, DataStorePickleMixin):
+    """Store for reading and writing data via zarr
+    """
+
+    # need some special secret attributes to tell us the dimensions
+    _DIMENSION_KEY = '_ARRAY_DIMENSIONS'
+
+    def __init__(self, store=None, mode='a', synchronizer=None, group=None,
+                    auto_chunk=True, writer=None, autoclose=None):
+        opener = functools.partial(_open_zarr_group, store, mode,
+                                   synchronizer, group)
+        self.ds = opener()
+
+        self._mode = mode
+        self._synchronizer = synchronizer
+        self._group = group
+        self._auto_chunk = auto_chunk
+
+        # zarr stores don't need to be opened, closed, or synced.
+        # So what do we do with all this logical about openers?
+        if autoclose:
+            raise NotImplementedError('autoclose=True is not implemented '
+                                      'for the zarr backend')
+        self._autoclose = False
+        self._isopen = True
+        self._opener = None
+
+        # initialize hidden dimension attribute
+        if self._DIMENSION_KEY not in self.ds.attrs:
+            self.ds.attrs[self._DIMENSION_KEY] = {}
+
+        # do we need to define attributes for all of the opener keyword args?
+        super(ZarrStore, self).__init__(writer)
+
+    def open_store_variable(self, name, zarr_array):
+        # I don't see why it is necessary to wrap self.ds[name]
+        # zarr seems to implement the required ndarray interface
+        # TODO: possibly wrap zarr array in dask with aligned chunks
+        data = indexing.LazilyIndexedArray(ZarrArrayWrapper(name, self))
+        dimensions, attributes = _get_zarr_dims_and_attrs(
+                                    zarr_array, self._DIMENSION_KEY)
+        encoding = {'chunks': zarr_array.chunks,
+                    'compressor': zarr_array.compressor,
+                    'filters': zarr_array.filters,
+                    'fill_value': zarr_array.fill_value}
+
+        var = Variable(dimensions, data, attributes, encoding)
+
+        if self._auto_chunk:
+            from dask.base import tokenize
+            # is this token enough?
+            token = tokenize(zarr_array)
+            name = 'zarr_array-%s' % token
+            # do we need to worry about the zarr synchronizer / dask lock?
+            lock = self._synchronizer
+            print("Chunking variable")
+            var = var.chunk(chunks=zarr_array.chunks, name=name, lock=lock)
```

Comment:
This does not work. Or rather, it does works but something downstream screws it up and loads the data. I don't understand what's happening between here and my `test_auto_chunk` test.